File: schneider_kalf/src/form/nodes/visitors/ExpressionVisitor.ts
import Addition from "../expressions/arithmetic/Addition";
import NumberLiteral from "../expressions/arithmetic/NumberLiteral";
import Expression from "../expressions/Expression";
import Multiplication from "../expressions/arithmetic/Multiplication";
import Or from "../expressions/boolean_expressions/Or";
import And from "../expressions/boolean_expressions/And";
import Negation from "../expressions/boolean_expressions/Negation";
import Variable from "../expressions/Variable";
import Division from "../expressions/arithmetic/Division";

interface ExpressionVisitor {
  visitAddition(addition: Addition): any;

  visitLiteral(literal: NumberLiteral): any;

  visitExpression(expression: Expression): any;

  visitMultiplication(multiplication: Multiplication): any;

  visitOr(or: Or): any;

  visitAnd(and: And): any;

  visitNegation(negation: Negation): any;

  visitVariable(variable: Variable): any;

  visitDivision(division: Division): any;
}

export default ExpressionVisitor;

File: nitpickers/pyql/pyql/ast/expression_visitor.py
from abc import abstractmethod


class ExpressionVisitor:

    @abstractmethod
    def visit_expression(self, expression):
        pass

    @abstractmethod
    def visit_identifier(self, identifier):
        pass

    @abstractmethod
    def visit_unary_expression(self, unary_expression):
        pass

    @abstractmethod
    def visit_binary_expression(self, binary_expression):
        pass

    @abstractmethod
    def visit_multiplication(self, multiplication):
        pass

    @abstractmethod
    def visit_division(self, division):
        pass

    @abstractmethod
    def visit_addition(self, addition):
        pass

    @abstractmethod
    def visit_subtraction(self, subtraction):
        pass

    @abstractmethod
    def visit_greater_than(self, greater_than):
        pass

    @abstractmethod
    def visit_less_than(self, less_than):
        pass

    @abstractmethod
    def visit_greater_than_or_equal(self, greater_than_or_equal):
        pass

    @abstractmethod
    def visit_less_than_or_equal(self, less_than_or_equal):
        pass

    @abstractmethod
    def visit_equals(self, equals):
        pass

    @abstractmethod
    def visit_not_equals(self, not_equals):
        pass

    @abstractmethod
    def visit_and(self, and_expression):
        pass

    @abstractmethod
    def visit_or(self, or_expression):
        pass

    @abstractmethod
    def visit_not(self, not_unary_expression):
        pass

    @abstractmethod
    def visit_literal(self, literal):
        pass

    @abstractmethod
    def visit_string_literal(self, string_literal):
        pass

    @abstractmethod
    def visit_integer_literal(self, int_literal):
        pass

    @abstractmethod
    def visit_decimal_literal(self, decimal_literal):
        pass

    @abstractmethod
    def visit_boolean_literal(self, bool_literal):
        pass

    @abstractmethod
    def visit_money_literal(self, money_literal):
        pass
